### ISSUE TITLE
[data grid] Implement Pivoting

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-x#214]
# What is a pivot table?

#### TLDR
_Transform row data from given selected field(s) into column headers; values under these columns are aggregations of all the rows that share the same value on the selected fields(s)_

A pivot table is a statistics tool that summarizes and reorganizes selected columns and rows of data in a spreadsheet or data table to obtain a desired report.

It’s an extension of the behavior of combining row grouping and aggregation.
With the ability to pivot columns, users can break down the row-grouped values in a new dimension, and the aggregation functions are used to compute the crossed values between rows and columns. 

It’s a tool that targets end-users and enables them to build a new table by combining different computed values based on the original dataset.

# How does pivoting work?

A pivot table is built over four distinct concepts.

- **Rows**: When a field is chosen to provide the rows, it populates the first column with the unique values of that field. It’s the logic of `Row grouping`.
- **Columns**: When a `field` is set for the columns, the unique values of that field are listed across the top as column headers. It is similar to `Row grouping`, but with columns.
This property is what effectively gives the sense of data pivoting.
- **Computed values:** When a field is chosen, its values are aggregated based on the cross between rows and columns established by the previous inputs. The most common aggregation are sum, average, minimum and maximum. 
It’s the logic from `Aggregation`, but extended to aggregate the value in multiple columns represented by a field’s unique values.
- **Filters:** Filters original dataset.

### Examples of use cases

- [https://www.youtube.com/watch?v=-LM6-BJd1HA](https://www.youtube.com/watch?v=-LM6-BJd1HA)
- Extracting the sum of gold medals per country each year:

    Original dataset    
    ![Original dataset](https://user-images.githubusercontent.com/550141/198119151-df52fd03-4724-412e-810b-f57d8e875acb.png)

    Row group: Country, 
    Columns: Year, 
    Computed Values: Gold (sum)
    ![Row group: Country, 
    Columns: Year, 
    Computed Values: sum(Gold)](https://user-images.githubusercontent.com/550141/198119219-864496ad-6329-44c4-98a5-1d2ac7421628.png)

     Row group: Year, 
    Columns: Country, 
    Computed Values: Gold (sum)
    ![Row group: Year, 
    Columns: Country, 
    Computed Values: Gold](https://user-images.githubusercontent.com/550141/198119464-090efaaf-3e9f-4ca3-be56-a04a8602e32d.png)

When a user turns on the pivoting:

1. The Data Grid stops displaying the original dataset
2. The Data Grid displays the tool for building the pivot table:
    - A list of all the available fields.
    - 3 inputs for columns, rows, and computed values.
    - Access to the filter panel to apply filters to the original dataset
3. The Data Grid renders the pivot table based on which fields are used on each of the four inputs.

# Where can users access this function?

At the new “Manage columns” panel
[https://github.com/mui/mui-x/issues/5700](https://github.com/mui/mui-x/issues/5700)

# Challenges

- UI/UX design of a pivoting panel
- Rendering autogenerated Columns (maybe Column definitions)
- Crossing Aggregation (Rows vs columns)

### Benchmark

- https://airtable.com/marketplace/blkyAnUVk7O0NMrQp/pivot-table
- https://help.tableau.com/current/prep/en-us/prep_pivot.htm
- https://www.ag-grid.com/javascript-grid-pivoting/
- https://ej2.syncfusion.com/react/demos/#/material/pivot-table/default
- https://demos.telerik.com/kendo-ui/pivotgrid/index
- https://demos.telerik.com/kendo-ui/pivotgridv2/index
- https://webix.com/pivot/
- https://www.htmlelements.com/docs/pivot-table/
- https://js.devexpress.com/Demos/WidgetsGallery/Demo/PivotGrid/Overview/jQuery/Light/
- Excel 

<img width="122" alt="Screenshot 2022-05-03 at 14 34 27" src="https://user-images.githubusercontent.com/3165635/166453616-c46d8019-4263-4d6b-993e-782abeb9cf52.png">

**Comment by itosi** [comment]
im looking foward to it :D

**Comment by LukeXF** [comment]
Is this still due for Q4? Hoping to see it soon!

**Comment by shubhamgoswamiofficial** [comment]
I thought we are the only one who always extend the deadline :laughing:

**Comment by dagi377** [comment]
Is this feature coming out by the end of this year?

**Comment by akshayknz** [comment]
This feature will come out in ....

**Comment by oliviertassinari** [comment]
We don't yet have a deadline for this feature. We have changed the roadmap presentation a bit. We now have an internal roadmap where we set ambitious targets and a public one where we try to under promise, so we can more consistently overdeliver.

**Comment by KlausFabian** [comment]
Main issue for me is that I can't make column headers stick - not being scrolled out. 
Maybe there is a way but I didn't see it.

**Comment by m4theushw** [comment]
@KlausFabian Is this what you're looking for? https://mui.com/components/data-grid/columns/#column-pinning

**Comment by KlausFabian** [comment]
@m4theushw That looks exactly what I have been looking for!!! Thanks a lot.

**Comment by rmrice** [comment]
Hoping to see this re-prioritized - my team will definitely use this as soon as it's available.

**Comment by GJmelody** [comment]
Looking forward to this new feature! Hope it can be prioritized

**Comment by paulleonartcalvo** [comment]
As a customer we would like to see this re-prioritized

**Comment by srinivaassunil** [comment]
We are in need of this feature very much, Please re-prioritize this.

**Comment by joserodolfofreitas** [comment]
I'm happy to listen to your pain points, expectations, and use cases. 
Let us know if you're willing to [give us a user interview](https://forms.gle/q9wqWtXUvR2kmzPA6).

**Comment by Rvalle23** [comment]
Hi @joserodolfofreitas ; thats an amazing initial point; we would love yo see that; let us know and in the future if any testing needed. Also; probably would connect soon; there were some issues with the export excel functionality. Kind regards, Ricardo

**Comment by davidespejo** [comment]
We really need this feature to make datagrid even more powerful!

**Comment by thiagoaragao** [comment]
That's killer for us!

**Comment by aimad-majdou** [comment]
Any updates on this ?

**Comment by demetriad** [comment]
Second that. I feel like this feature is a must.

**Comment by imhazige** [comment]
+1

**Comment by ks0430** [comment]
+1

**Comment by vaynevayne** [comment]
Excuse me, this feature requires the server to return all the data, right?

**Comment by h4r3en** [comment]
Can anyone please please provide ETA for this

**Comment by vaynevayne** [comment]
> Can anyone please please provide ETA for this

I do not look forward to, can not afford, Ha Ha

**Comment by olegschwarz** [comment]
Really need this asap!
We're on the pro now, we will upgrade to premium if this will be released soon!

**Comment by mrityunjay-cnuro** [comment]
+1 for this feature

**Comment by andrejremic** [comment]
+1 - We would also need it in near future, otherwise we will have to find alternative solution.

**Comment by srgg** [comment]
With all the respect, is there any progress on that?

**Comment by oliviertassinari** [comment]
> Is there any progress on this?

@srgg We recently moved the feature to "Next in queue" in our public MUI X roadmap https://github.com/mui/mui-x/projects/1#card-45123122

<img width="745" alt="Screenshot 2023-07-15 at 18 05 20" src="https://github.com/mui/mui-x/assets/3165635/856bf5d5-23a5-49b0-9d55-6dd27970b9f3">

Our next step is to run a proof of concept. 

> Can anyone please please provide ETA for this

@h4r3en Assuming we plan for the worse case: a stable API release for March 2024, and the first unstable API release for Q4 2023 feel like something we could pull off.

**Comment by voidnothings** [comment]
+1

**Comment by scamden** [comment]
> the first unstable API release before the end of year

did this happen? can i test it somewhere?

**Comment by syuan100** [comment]
> > the first unstable API release before the end of year
> 
> did this happen? can i test it somewhere?

Would also like to know if unstable API is available somewhere

**Comment by olegschwarz** [comment]
Same here, Where can we check the existing API for pivoting? Is it available somewhere?

**Comment by cherniavskii** [comment]
Hi everyone,
We plan to release the Pivoting feature this quarter!
Here is a proof of concept PR: https://github.com/mui/mui-x/pull/9877.
You can preview the feature here https://deploy-preview-9877--material-ui-x.netlify.app/x/react-data-grid/pivoting/ - give it a try.
The UI for pivoting will be different, but the overall concept will stay the same.
Let us know what you think!

**Comment by srinivaassunil** [comment]
If you are looking for an alternative meanwhile, i have found this library
https://perspective.finos.org/

**Comment by delpiersos** [comment]
Is this still on track for this quarter?

**Comment by cherniavskii** [comment]
Hi @delpiersos 
We didn't make it in Q1, but it's on track for Q2.
You can check out the preview in https://github.com/mui/mui-x/pull/9877: https://deploy-preview-9877--material-ui-x.netlify.app/x/react-data-grid/pivoting/
We would appreciate your feedback!

**Comment by delpiersos** [comment]
> Hi @delpiersos We didn't make it in Q1, but it's on track for Q2. You can check out the preview in #9877: https://deploy-preview-9877--material-ui-x.netlify.app/x/react-data-grid/pivoting/ We would appreciate your feedback!

Hi @cherniavskii - thanks for the update, really hoping to see it soon. 
Some feedback on the preview you've shared (not sure if there's anything more recent); note i'm a PM, not an engineer - apologies if my views appear to take things too simply or I'm missing a technical point somewhere:

- When opening the "Columns" or "Filters" menus in the top - columns get repeated - seemingly to n (i.e. the matching the number of occurrences across the data set). Really, these data elements are common entity across all entries, so by removing "Gross", it should be removing it from all pivot columns.
- Filters similarly should only contain each unique item in that way, when I filter on "Gross", I'm expecting to just pick Gross and my filter, and show items that match, regardless of the Year group or whatever it might be. I should also be able to filter on "Year" (the column) in this example, but it's not available.
<img width="800" alt="Screenshot 2024-05-08 at 1 39 18 PM" src="https://github.com/mui/mui-x/assets/28945021/0067808c-b386-4207-af23-47b4328cc4b1">
- Given that the Values contain items that should allow mathematical operators - i.e. sum of Gross, avg of rating; I would expect the filters to allow for numeric filtering like the below, but perhaps the data set is poorly defined?:
-- greater than, greater than or equal to
-- less than, less than or equal to
-- between

- Let me hide the pivot controls to take back screen real estate to view the results

It's really encouraging progress; the feedback above make it a bit difficult to incorporate into my use case in a way that my users would be able to leverage.

**Comment by RomanMain** [comment]
Hey. @cherniavskii Can we know the approximate delivery time for this functionality in the stable version? Since Q2 is over, when can we expect it? Thank you!

**Comment by anu365** [comment]
Hey. @cherniavskii Can we know the approximate delivery time for this functionality in the stable version? Since Q3 is also over, when can we expect it?

**Comment by oliviertassinari** [comment]
> Can we know the approximate delivery time for this functionality in the stable version?

@anu365 To connect the dots, in https://mui.com/blog/mui-x-v8-alpha-zero/ & https://x.com/MUI_hq/status/1861781032933953966 we communicated that this issue is making it for MUI X v8.0.

**Comment by alexkubica** [comment]
upvote!

**Comment by daniellepaquetteharvey** [comment]
@oliviertassinari I have premium support. I can raise an issue if it helps having this feature move forward.

**Comment by RobertoPrivat** [comment]
We would use pivoting for ad analytics. Facebook has a pivot functionality in their ad reporting tool which is very helpful to quickly understand the impact of advertisement configurations. We enrich Facebook data with other non-platform data and thus, would like to replicate the pivot function in our front-end, if it is available.

**Comment by joey-io** [comment]
We would love a pivot table :)

**Comment by cherniavskii** [comment]
For those interested in the feature – we're going to release it soon!
You can check out the docs preview at https://deploy-preview-9877--material-ui-x.netlify.app/x/react-data-grid/pivoting/

**Comment by cherniavskii** [comment]
> We have some non-aggregatable measures, for which we just send the total values in the data.

@astragier I'm not sure I understand. By "non-aggregatable measures," do you mean non-numeric values? And by "total values in the data," do you mean the count of values?
If my assumptions above are correct, there's [`size` aggregation function ](https://mui.com/x/react-data-grid/aggregation/#built-in-functions) that can be used with non-numeric values.

**Comment by astragier** [comment]
> > We have some non-aggregatable measures, for which we just send the total values in the data.
> 
> [@astragier](https://github.com/astragier) I'm not sure I understand. By "non-aggregatable measures," do you mean non-numeric values? And by "total values in the data," do you mean the count of values? If my assumptions above are correct, there's [`size` aggregation function ](https://mui.com/x/react-data-grid/aggregation/#built-in-functions) that can be used with non-numeric values.

Here is my measure definition:
`Price index = SUM[product, store]( sales * price / avg(competitor price) ) / SUM[product, store]( sales )`
It is numeric but non aggregable as turnover for instance. Total price index cannot be expressed thanks to an agregation of the price indexes of all the product brands for instance.
Here is my pivot table as an example:
![Image](https://github.com/user-attachments/assets/20675a78-93c3-492e-a39c-f4bf66a7e4f9)
In that case, my pivot table won't compute the pivot itself but I give it in the input data:
```
{
    "cells": [
        {
            "price_index_final": 98.0682102245562
        },
        {
            "brand_type": "MDD",
            "price_index_final": 111.577728334608
        },
        {
            "brand": "AIX BAINS",
            "brand_type": "MDD",
            "price_index_final": 145.16528772415825
        },
        {
            "brand": "CLAIRPASTI",
            "brand_type": "MDD",
            "price_index_final": 96.90827914371435
        },
        {
            "brand": "LOOK",
            "brand_type": "MDD",
            "price_index_final": 104.25540538294041
        },
        {
            "brand": "MAGNES.",
            "brand_type": "MDD",
            "price_index_final": 111.91110585231247
        },
        {
            "brand": "NATIVE",
            "brand_type": "MDD",
            "price_index_final": 111.52125059526857
        },
        {
            "brand": "ONDINE",
            "brand_type": "MDD",
            "price_index_final": 100.18958939861406
        },
        {
            "brand": "PAQUITO",
            "brand_type": "MDD",
            "price_index_final": 139.2977177860899
        },
        {
            "brand": "PASTOURET",
            "brand_type": "MDD",
            "price_index_final": 102.51942500769107
        },
        {
            "brand_type": "MN",
            "price_index_final": 96.07905369889461
        },
        {
            "brand": "1212",
            "brand_type": "MN",
            "price_index_final": 106.5245565440199
        },
        {
            "brand": "ARIZONA",
            "brand_type": "MN",
            "price_index_final": 105.24155642156359
        },
        {
            "brand": "BERGER",
            "brand_type": "MN",
            "price_index_final": 104.17136012554924
        },
        {
            "brand": "CALYPSO",
            "brand_type": "MN",
            "price_index_final": 103.46222137601855
        },
        {
            "brand": "CAPRI SUN",
            "brand_type": "MN",
            "price_index_final": 94.34844223176209
        },
        {
            "brand": "CASANIS",
            "brand_type": "MN",
            "price_index_final": 106.86527800726756
        },
        {
            "brand": "COCA",
            "brand_type": "MN",
            "price_index_final": 96.03821815583962
        },
        {
            "brand": "CONTREX",
            "brand_type": "MN",
            "price_index_final": 101.58246992954031
        },
        {
            "brand": "COURMAYEUR",
            "brand_type": "MN",
            "price_index_final": 101.12220626252558
        },
        {
            "brand": "CRISTALINE",
            "brand_type": "MN",
            "price_index_final": 93.0803375897801
        },
        {
            "brand": "CUSENIER",
            "brand_type": "MN",
            "price_index_final": 105.77259475218659
        },
        {
            "brand": "DR PEPPER",
            "brand_type": "MN",
            "price_index_final": 101.33974588623055
        },
        {
            "brand": "DUVAL",
            "brand_type": "MN",
            "price_index_final": 93.92906070414845
        },
        {
            "brand": "EVIAN",
            "brand_type": "MN",
            "price_index_final": 96.77390350316533
        },
        {
            "brand": "FER.BRANCA",
            "brand_type": "MN",
            "price_index_final": 107.00661736084078
        },
        {
            "brand": "FLORANIS",
            "brand_type": "MN",
            "price_index_final": 103.35710458048227
        },
        {
            "brand": "FRUI.SHOOT",
            "brand_type": "MN",
            "price_index_final": 102.77271069988757
        },
        {
            "brand": "HEPAR",
            "brand_type": "MN",
            "price_index_final": 96.56387573903977
        },
        {
            "brand": "HEROIC SPT",
            "brand_type": "MN",
            "price_index_final": 100.94633702479723
        },
        {
            "brand": "JANOT",
            "brand_type": "MN",
            "price_index_final": 115.01809408926418
        },
        {
            "brand": "LIPTON",
            "brand_type": "MN",
            "price_index_final": 98.02826148555759
        },
        {
            "brand": "MONT BLANC",
            "brand_type": "MN",
            "price_index_final": 98.54585821634733
        },
        {
            "brand": "MT ROUCOUS",
            "brand_type": "MN",
            "price_index_final": 92.5807387810962
        },
        {
            "brand": "NARUTO",
            "brand_type": "MN",
            "price_index_final": 108.86002886002886
        },
        {
            "brand": "OASIS",
            "brand_type": "MN",
            "price_index_final": 100.8939343823006
        },
        {
            "brand": "PASTIS 51",
            "brand_type": "MN",
            "price_index_final": 97.19024029875952
        },
        {
            "brand": "PAT PATROU",
            "brand_type": "MN",
            "price_index_final": 125.93737920371086
        },
        {
            "brand": "PEPSI",
            "brand_type": "MN",
            "price_index_final": 100.3327178842652
        },
        {
            "brand": "PICON",
            "brand_type": "MN",
            "price_index_final": 105.77475335189133
        },
        {
            "brand": "PULCO",
            "brand_type": "MN",
            "price_index_final": 102.60946117165146
        },
        {
            "brand": "RICARD",
            "brand_type": "MN",
            "price_index_final": 96.30031049692731
        },
        {
            "brand": "ST ANTONIN",
            "brand_type": "MN",
            "price_index_final": 97.93015460674845
        },
        {
            "brand": "SUZE",
            "brand_type": "MN",
            "price_index_final": 104.15704900870746
        },
        {
            "brand": "THONON",
            "brand_type": "MN",
            "price_index_final": 100.37677561012227
        },
        {
            "brand": "TROPICANA",
            "brand_type": "MN",
            "price_index_final": 104.58249077166019
        },
        {
            "brand": "VITTEL",
            "brand_type": "MN",
            "price_index_final": 101.66773470633173
        },
        {
            "brand": "VOLVIC",
            "brand_type": "MN",
            "price_index_final": 96.36249096239541
        },
        {
            "brand_type": "PP",
            "price_index_final": 110.16379948832736
        },
        {
            "brand": "1ER PRIX",
            "brand_type": "PP",
            "price_index_final": 97.78709943632852
        },
        {
            "brand": "No Name",
            "brand_type": "PP",
            "price_index_final": 111.48067308878703
        },
        {
            "brand": "TOP BUDGET",
            "brand_type": "PP",
            "price_index_final": 109.0855634818647
        }
    ],
    "rows": [
        "brand_type",
        "brand"
    ],
    "columns": [],
    "values": [
        "price_index_final"
    ],
}
```

**Comment by cherniavskii** [comment]
@astragier Interesting use case!
The client-side pivoting transforms the data, groups the rows, aggregates the values, and creates column groups.
Judging by the video you provided, you're looking for the server-side pivoting feature in which row groups and cell values are generated on the server.
We're focusing on delivering client-side pivoting first. But I will keep this use case in mind.

**Comment by cherniavskii** [comment]
@astragier By the way, is the pivot configuration panel part of the pivot table you're using, or is it a custom-built one?

**Comment by cherniavskii** [comment]
Closed by https://github.com/mui/mui-x/pull/9877
